test/compress/template-string.js

```diff
@@ -0,0 +1,316 @@
+        var bazbaz = `This is ${1/0}`;
+        var barbar = `This is NaN`;
+        var barbar = "This is ${0/0}";
+        var barber = 'This is ${0/0}';
+
+        var a = `4194304`;
+        var b = `16777216`; // Potential for further concatentation
+        var c = `${4**14}`; // Not worth converting
+    }
+}
+
+template_string_evaluate_with_many_segments: {
+    options = {
+        evaluate: true
+    }
+    beautify = {
+        quote_style: 3
+    }
+    input: {
+        var foo = `Hello ${guest()}, welcome to ${location()}${"."}`;
+        var bar = `${1}${2}${3}${4}${5}${6}${7}${8}${9}${0}`;
+        var baz = `${foobar()}${foobar()}${foobar()}${foobar()}`;
+        var buzz = `${1}${foobar()}${2}${foobar()}${3}${foobar()}`;
+    }
+    expect: {
+        var foo = `Hello ${guest()}, welcome to ${location()}.`;
+        var bar = `1234567890`;
+        var baz = `${foobar()}${foobar()}${foobar()}${foobar()}`;
+        var buzz = `1${foobar()}2${foobar()}3${foobar()}`;
+    }
+}
+
+template_string_with_many_segments: {
+    beautify = {
+        quote_style: 3
+    }
+    input: {
+        var foo = `Hello ${guest()}, welcome to ${location()}${"."}`;
+        var bar = `${1}${2}${3}${4}${5}${6}${7}${8}${9}${0}`;
+        var baz = `${foobar()}${foobar()}${foobar()}${foobar()}`;
+        var buzz = `${1}${foobar()}${2}${foobar()}${3}${foobar()}`;
+    }
+    expect: {
+        var foo = `Hello ${guest()}, welcome to ${location()}${"."}`;
+        var bar = `${1}${2}${3}${4}${5}${6}${7}${8}${9}${0}`;
+        var baz = `${foobar()}${foobar()}${foobar()}${foobar()}`;
+        var buzz = `${1}${foobar()}${2}${foobar()}${3}${foobar()}`;
+    }
+}
+
+template_string_to_normal_string: {
+    options = {
+        evaluate: true
+    }
+    beautify = {
+        quote_style: 0
+    }
+    input: {
+        var foo = `This is ${undefined}`;
+        var bar = "Decimals " + `${1}${2}${3}${4}${5}${6}${7}${8}${9}${0}`;
+    }
+    expect: {
+        var foo = "This is undefined";
+        var bar = "Decimals 1234567890";
+    }
+}
+
+template_concattenating_string: {
+    options = {
+        evaluate: true
+    }
+    beautify = {
+        quote_style: 3 // Yes, keep quotes
+    }
+    input: {
+        var foo = "Have a nice " + `day. ${`day. ` + `day.`}`;
+        var bar = "Have a nice " + `${day()}`;
+    }
+    expect: {
+        var foo = "Have a nice day. day. day.";
+        var bar = "Have a nice " + `${day()}`;
+    }
+}
+
+evaluate_nested_templates: {
+    options = {
+        evaluate: true
+    }
+    beautify = {
+        quote_style: 0
+    }
+    input: {
+        var baz = `${`${`${`foo`}`}`}`;
+    }
+    expect: {
+        var baz = "foo";
+    }
+}
+
+enforce_double_quotes: {
+    beautify = {
+        quote_style: 1
+    }
+    input: {
+        var foo = `Hello world`;
+        var bar = `Hello ${'world'}`;
+        var baz = `Hello ${world()}`;
+    }
+    expect: {
+        var foo = "Hello world";
```

Comment:
Must use `expect_exact` in **all** tests because quotes in regular strings in `expect` blocks can be either single or double and still be successful.